### ISSUE TITLE
Shell bugs on different systems

### ISSUE THREAD

**Issue by vit9696** [acidanthera/bugtracker#364]
There are multiple shell issues including, some of which have dedicated bugreports, some of which have patches upstreamed, or just nothing. This issue will be the parent issue for all separate ones

- [x] Missing device path, as done by tool launch, leads to assertion in debug builds (#362).
- [x] Printing `\t` on some firmwares in UEFI Shell makes screen output entirely disappear afterwards. Reported on Lenovo Yoga Ivy laptop by @usr-sse2.
- [x] In editor Ctrl + E does not work SimpleTextInputProtocolGuid. Try implementing UDK Hunk SimpleTextInput2 over SimpleTextInput? Explore it? Reported on Lenovo Yoga Ivy laptop by @usr-sse2.
- [x] On legacy EDK-based Macs using load/drivers commands results in errors.
- [x] Various patches from https://github.com/acidanthera/OpenCoreShell/tree/v1.0.1/Patches. We should document them more properly.
- [x] In editor Shift+letter doesn’t work to produce uppercase letters with USB keyboards at all (on Dell Optiplex Ivy desktop) and works only since second time (Shift-letter-letter) with PS/2 keyboards (on Lenovo Yoga Ivy laptop). Reported by @usr-sse2.

**Comment by vit9696** [comment]
So, after upgrading our EDK II base I noticed new problems with Shell on MacBookPro12,1:

- Upstream started to use CreateEventEx even further, which led me creating [another patch](https://github.com/acidanthera/OpenCoreShell/blob/master/Patches/debug-lib-compat.diff) in addition to previously removed CreateEventEx in [hii patch](https://github.com/acidanthera/OpenCoreShell/blob/master/Patches/hii-database.diff).
- MacBookPro12,1 firmware has no SimpleTextInputEx protocol, so I get `No SimpleTextInputEx was found. CTRL-based features are not usable.` warning, and `edit` command does not work for me either.

I wonder whether the correct solution to these issues is to create a lib that will implement its own copy of `EFI_BOOT_SERVICES`. This way we can return pointers to Hii protocols by customising Locate/Handle protocol functions (for e.g. Hii) and make a dummy CreateEventEx function. Perhaps this is more error-prone than trying to fill EDK II with more patches.

**Comment by vit9696** [comment]
I am not positive it is worth implementing SimpleTextInputEx protocol for legacy macs, and the rest is fixed in master. Closing.

**Comment by startergo** [comment]
On cMP the shell fails with `No SimpleTextInputEx was found. CTRL-based features are not usable. `.
Booting through Apple bootpicker with compatible EFI vbios card to a supported shell like the one from rEFIt works very well with the exception of the 640x480 resolution. Same behavior is observed if from Apple boot screen the system is booted to rEFIt. Files can be edited in shell and the CTRL-based functions are usable. If the same rEFIt-->shell.efi is loaded from OC it does not work.

**Comment by vit9696** [comment]
Who does install SimpleTextInputEx when it works? I do not think MacPro firmware supports SimpleTextInputEx at all.

**Comment by startergo** [comment]
> UPD: _to a supported shell_ is the key. If you use legacy EFI Shell that does not use SimpleTextInputEx, like the one from rEFIt, it will work everywhere.

Right, I expected it to work everywhere, but it only works when initialized directly from the Apple Boot selector and not from OC.

**Comment by vit9696** [comment]
This is because you did not switch to text mode, which is done by OpenShell at startup and other compatible tools. You could run legacy EFI Shell from OpenShell, and it will work.

**Comment by startergo** [comment]
> This is because you did not switch to text mode
And how do I do that? `Picker mode Builtin`?

**Comment by vit9696** [comment]
> You could run legacy EFI Shell from OpenShell, and it will work.

**Comment by startergo** [comment]
I have tested extensively based on suggestions, @dakanji also tried incorporating the module:
```
Tested Refit (EFI v1) , EDK (EFI v1 - They have a v2 but didn't bother with that) and OpenCore (EFI v2) Versions.

All work with the OpenCore version showing the error about SimpletextInputEX. The other two don't show this message but might just be that the OpenCore version is compiled to show it.

When I say work, I mean they can be used but have not gone as far as testing the various F* commands.

I did test the suggestion about compiling that code vit9696 suggested into the Refit version and positive it will not make a difference.

First, I was able to confirm that both the EDK and OpenCore Versions, change the console mode to text by themselves. while the Refit version does not. So, adding the the code would appear to make sense on the surface.

However, Refind definitely changes the mode by itself using code that is the same as the suggested one.

Doesn't seem what ever is the issue is not going to be solved by simply recompiling with that code (See Caveat).

As the Refit version does not have this change code however, the EDK v1 one seems to the best to go with since it already seems to have this code in and is EFI v1 (equivalent to Refit). I have actually added it to the MyBootMgr Chainloader.

CAVEAT: I didn't actually test whether those commands work. I just didn't see the writing at the bottom as in the image you sent earlier but I suppose this doesn't mean they are not available.

Will figure out how exactly to test them. I did press F7 once and got a reaction (new line).

```

**Comment by startergo** [comment]
I got some form of success, but only that it will not hang the system after which I had to reboot the machine. Now at least it gives me an error on an attempt to edit any text file:
![image](https://user-images.githubusercontent.com/7897244/89598617-3f92cc80-d82b-11ea-821e-e4ae4347b983.png)
What I did was I compiled from UDK2018v2 the MdeModule package and loaded `GraphicsConsoleDxe.efi` on boot.

**Comment by vit9696** [comment]
@startergo look, I have already explained clearly what is going on here and asked you some direct questions rather than requested for experiments.

* UEFI Shell relies on SimpleInputEx protocol, which is available only on modern firmwares. Macs before Skylake do not have it. OpenCore (and UDK, and edk2) use UEFI Shell in some form.
* EFI Shell only uses SimpleInput protocol, which is available everywhere, including your MacPro. It is not as feature rich as it does not allow catching e.g. `Ctrl+KEY` combinations, which modern `edit` commands relies on.
* It is rather time consuming and in fact a bit complicated to implement `SimpleInputEx` protocol in a reasonable way over `SimpleInput` or `AppleKeyMapAggregator`, but I filed an issue [here](https://github.com/acidanthera/bugtracker/issues/1080), so that whoever wants to do it has a way to track it. I do not believe we will do it ourselves though.
* Launching any program from OpenCore (e.g. as a Tool) will make OpenCore switch to `graphics mode`. Which means that any written text will not be visible onscreen. This behaviour is what Apple EFI BootPicker implements as well. To get any text rendered the application needs a way to switch to `text mode`. Currently all OpenCore utilities, including our UEFI Shell, are compiled with `OcConsoleControlEntryModeLib` code, which does exactly that.  Your EFI Shell from rEFIt does not do that, and as a result you will always see black screen after launching it from OpenCore.
* To run a tool that does not switch to `text mode` (e.g. legacy EFI Shell) from OpenCore your current only option is to run OpenCore UEFI Shell, and then run the legacy EFI Shell in it. In future we might let specifying the console mode via `config.plist`. I also filed an issue [here](https://github.com/acidanthera/bugtracker/issues/1081).
